File: examples/authenticated-client.js
import { GoogleAuth } from 'google-auth-library';
import axios from 'axios';
import fs from 'fs';
import path from 'path';

/**
 * Example of authenticating to the Cloud Run service using a service account
 */
async function authenticatedRequest() {
  // Determine the service account key path
  // Supports both: SERVICE_ACCOUNT_PATH env var, project-specific keys, or default symlink
  let keyFilename = process.env.SERVICE_ACCOUNT_PATH;

  if (!keyFilename) {
    // Try to read project ID from .env for project-specific key
    if (fs.existsSync('.env')) {
      const envContent = fs.readFileSync('.env', 'utf8');
      const projectIdMatch = envContent.match(/^PROJECT_ID=(.+)$/m);
      if (projectIdMatch) {
        const projectId = projectIdMatch[1].trim();
        const projectSpecificPath = `service_accounts/${projectId}_service_account.json`;
        if (fs.existsSync(projectSpecificPath)) {
          keyFilename = projectSpecificPath;
        }
      }
    }

    // Fall back to default symlink
    if (!keyFilename && fs.existsSync('service_account.json')) {
      keyFilename = 'service_account.json';
    }
  }

  if (!keyFilename) {
    console.error('Error: No service account key found.');
    console.error('Run ./scripts/download-service-account-key.sh to download it');
    process.exit(1);
  }

  console.log(`Using service account key: ${keyFilename}`);

  // Initialize the Google Auth client with the service account key
  const auth = new GoogleAuth({
    keyFilename
  });

  // Get your Cloud Run service URL (replace with your actual URL)
  const SERVICE_URL = process.env.SERVICE_URL || 'https://your-service-name-xxxxx.run.app';
  
  try {
    // Get an ID token for the service URL
    const client = await auth.getIdTokenClient(SERVICE_URL);
    const headers = await client.getRequestHeaders();

    // Make a request to your service
    const response = await axios.post(
      `${SERVICE_URL}/run`,
      {
        prompt: 'What is 5+5?',
        maxTurns: 1,
        mcpConfigJson: { mcpServers: {} },
        allowedTools: [],
        permissionMode: 'acceptEdits'
      },
      {
        headers: {
          ...headers,
          'Content-Type': 'application/json'
        }
      }
    );

    console.log('Response:', response.data);
  } catch (error) {
    console.error('Error:', error.message);
    if (error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response data:', error.response.data);
    }
  }
}

// Run the example
authenticatedRequest();

File: examples/authenticated-client.py
import os
import json
import requests
import re
from pathlib import Path
from google.auth import jwt
from google.auth.transport.requests import Request

def authenticated_request():
    """
    Example of authenticating to the Cloud Run service using a service account
    """
    # Determine the service account key path
    # Supports both: SERVICE_ACCOUNT_PATH env var, project-specific keys, or default symlink
    SERVICE_ACCOUNT_FILE = os.environ.get('SERVICE_ACCOUNT_PATH')

    if not SERVICE_ACCOUNT_FILE:
        # Try to read project ID from .env for project-specific key
        env_file = Path('.env')
        if env_file.exists():
            with open(env_file, 'r') as f:
                env_content = f.read()
                match = re.search(r'^PROJECT_ID=(.+)$', env_content, re.MULTILINE)
                if match:
                    project_id = match.group(1).strip()
                    project_specific_path = f'service_accounts/{project_id}_service_account.json'
                    if Path(project_specific_path).exists():
                        SERVICE_ACCOUNT_FILE = project_specific_path

        # Fall back to default symlink
        if not SERVICE_ACCOUNT_FILE and Path('service_account.json').exists():
            SERVICE_ACCOUNT_FILE = 'service_account.json'

    if not SERVICE_ACCOUNT_FILE:
        print('Error: No service account key found.')
        print('Run ./scripts/download-service-account-key.sh to download it')
        return

    print(f'Using service account key: {SERVICE_ACCOUNT_FILE}')
    
    # Your Cloud Run service URL (replace with your actual URL)
    SERVICE_URL = os.environ.get('SERVICE_URL', 'https://your-service-name-xxxxx.run.app')
    
    try:
        # Load the service account credentials
        with open(SERVICE_ACCOUNT_FILE, 'r') as f:
            service_account_info = json.load(f)
        
        # Create credentials object
        credentials = jwt.Credentials.from_service_account_info(
            service_account_info,
            audience=SERVICE_URL
        )
        
        # Refresh the credentials to get an ID token
        credentials.refresh(Request())
        
        # Make a request to your service
        headers = {
            'Authorization': f'Bearer {credentials.token}',
            'Content-Type': 'application/json'
        }
        
        data = {
            # Your request body here
            'command': 'echo "Hello from authenticated Python client"',
            'directory': '/tmp'
        }
        
        response = requests.post(
            f'{SERVICE_URL}/process',
            headers=headers,
            json=data
        )
        
        print('Response:', response.json())
        
    except FileNotFoundError:
        print(f'Error: Service account file {SERVICE_ACCOUNT_FILE} not found')
        print('Run ./scripts/download-service-account-key.sh to download it')
    except Exception as e:
        print(f'Error: {str(e)}')

if __name__ == '__main__':
    authenticated_request()